### PR TITLE
feat(swap): Allow cancellation of an open order

### PR DESCRIPTION
**Pull request by michalszorad**
Resolves [YOMO-750](https://emurgo.atlassian.net/jira/software/c/projects/YOMO/issues/YOMO-750)

Note this PR does not have HW wallet support yet.

[YOMO-750]: https://emurgo.atlassian.net/browse/YOMO-750?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by michalszorad** [comment]
~Please note it's impossible to place a limit order on mobile right now, as the limit orders are executed as if they are market orders.~ Limit orders placed through the web are also not possible to cancel, as confirmed on Slack, making it impossible to test.

EDIT: Limit orders are possible to place on mobile, but we're still however getting a signature error

```
{"error": {"response": "transaction submit error ShelleyTxValidationError ShelleyBasedEraBabbage (ApplyTxError [UtxowFailure (AlonzoInBabbageUtxowPredFailure (ShelleyInAlonzoUtxowPredFailure (MissingVKeyWitnessesUTXOW (fromList [KeyHash \"f218dd624987919f2cf899113c5e8650685e2e32268aac4064718f48\"])))),UtxowFailure (AlonzoInBabbageUtxowPredFailure (PPViewHashesDontMatch (SJust (SafeHash \"f243d69559fa63638d46f9ff5a105f01dd9d168506e05422004cd959f0c8b5c3\")) (SJust (SafeHash \"17dfdd4dfde26ba016c2c28c6395fa04b54c4cc0b6d122bd68eb0e9aaa685db5\"))))])"}}
```

I'm in touch with MuesliSwap to resolve the issue

File: apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/OpenOrders.tsx [hunk 1]
```diff
@@ -349,18 +491,71 @@ const ModalContent = ({
...
+  bech32Address: string
 }) => {
   const strings = useStrings()
+
+  const {order} = useSwap()
+  const wallet = useSelectedWallet()
+
+  const [fee, setFee] = useState<string | null>(null)
+
+  const getFee = useCallback(async () => {
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/OpenOrders.tsx]: wrap it within a react-query to handle the async state, and pass the dependencies over, leverage the suspense or isLoading, I'd recommend to extract most of this logic that is not UI, into a `/ListOrder/helpers.tsx` and have the hooks and most of the helpers over there.

Comment (michalszorad) [reply apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/OpenOrders.tsx]: Thank you, that's now updated.

File: apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/OpenOrders.tsx [hunk 2]
```diff
@@ -284,6 +370,68 @@ const HiddenInfo = ({
   )
 }
 
+const PasswordModal = ({onConfirm}: {onConfirm: (password: string) => Promise<void>}) => {
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/OpenOrders.tsx:372]: isn't it part of the confirmation with password screen?

File: apps/wallet-mobile/src/yoroi-wallets/cardano/types.ts [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import {WalletChecksum as WalletChecksumType} from '@emurgo/cip4-js'
 import * as CoreTypes from '@emurgo/cross-csl-core'
 import {BaseAddress} from '@emurgo/cross-csl-core'
+import {PrivateKey} from '@emurgo/csl-mobile-bridge'
```
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/types.ts]: isn't exposed in the core?

File: apps/wallet-mobile/src/yoroi-wallets/cardano/utils.ts [hunk 4]
```diff
@@ -305,3 +316,40 @@ export const selectFtOrThrow = (token: Balance.TokenInfo): Balance.TokenInfo =>
   }
   throw new Error(`Token type "${token.kind}" is not a fungible token`)
 }
+
+export const generateCIP30UtxoCbor = async (utxo: RawUtxo) => {
+  const txHash = await TransactionHash.from_bytes(Buffer.from(utxo.tx_hash, 'hex'))
+  if (!txHash) throw new Error('Invalid tx hash')
+
+  const index = BigNumber(utxo.utxo_id.split(':')[1]).toNumber()
```
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/utils.ts]: isn't the BigNum from here

File: apps/wallet-mobile/src/yoroi-wallets/cardano/utils.ts [hunk 5]
```diff
@@ -1,7 +1,18 @@
...
+import {
+  Address,
+  BigNum,
+  PrivateKey,
+  TransactionHash,
+  TransactionInput,
+  TransactionOutput,
+  TransactionUnspentOutput,
+  Value,
+} from '@emurgo/csl-mobile-bridge'
```
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/utils.ts]: this should be csl-core (bridge wasn't meant to be imported)

File: apps/wallet-mobile/src/yoroi-wallets/cardano/utils.ts (hunk 4 above)
Comment (lisicky) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/utils.ts]: index is integer

File: apps/wallet-mobile/package.json [hunk 6]
```diff
@@ -105,8 +105,8 @@
     "@cardano-foundation/ledgerjs-hw-app-cardano": "^6.0.0",
     "@emurgo/cip14-js": "^3.0.1",
     "@emurgo/cip4-js": "1.0.7",
-    "@emurgo/cross-csl-core": "^2.9.0",
-    "@emurgo/cross-csl-mobile": "^2.9.0",
+    "@emurgo/cross-csl-core": "file:/Users/michalszorad/Desktop/cross-csl/packages/cross-csl-core",
```
Comment (michalszorad) [inline apps/wallet-mobile/package.json]: This will need to be updated once new version is out

Comment (michalszorad) [reply apps/wallet-mobile/package.json]: Already updated

File: apps/wallet-mobile/__mocks__/@emurgo/csl-mobile-bridge.js
Comment (stackchain) [inline apps/wallet-mobile/__mocks__/@emurgo/csl-mobile-bridge.js:1]: is it sill needed after the changes?

File: apps/wallet-mobile/src/yoroi-wallets/cardano/utils.ts [hunk 7]
```diff
@@ -305,3 +306,40 @@ export const selectFtOrThrow = (token: Balance.TokenInfo): Balance.TokenInfo =>
...
+  if (!amount) throw new Error('Invalid amount')
+
+  const collateral = await CardanoMobile.Value.new(amount)
+  const output = await CardanoMobile.TransactionOutput.new(address, collateral)
+  const transactionUnspentOutput = await CardanoMobile.TransactionUnspentOutput.new(input, output)
+
+  return transactionUnspentOutput.toHex()
+}
+
+export const generateMuesliSwapSigningKey = async (rootKey: string, derivationPath: number[]) => {
```
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/utils.ts]:
```suggestion
export const createRawTxSigningKey = async (rootKey: string, derivationPath: number[]) => {
```

File: apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts [hunk 8]
```diff
@@ -453,6 +454,32 @@ export const makeShelleyWallet = (constants: typeof MAINNET | typeof TESTNET) =>
       return stakingKey
     }
 
+    public async signRawTx(txHex: string, pKeys: PrivateKey[]) {
```
Comment (lisicky) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts:457]: Do we need implement this function in two places ? Can we move it into separate file ?

Comment (michalszorad) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts:457]: I've created signatureUtils.ts file, extracted the logic and reused it in Shelley Wallet and Byron Wallet

File: apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/signatureUtils.ts
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/signatureUtils.ts:1]: if it's used by `shelley` + `byron` move into `common`

File: apps/wallet-mobile/__mocks__/@emurgo/csl-mobile-bridge.js
Comment (stackchain) [inline apps/wallet-mobile/__mocks__/@emurgo/csl-mobile-bridge.js:1]: jest mocks by __mocks__ follow the file name, but the export is different, I think we can delete it.

File: apps/wallet-mobile/src/features/Swap/common/ConfirmWithSpendingPassword/ConfirmWithSpendingPassword.tsx [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        autoComplete="off"
+      />
+
+      {error != null ? (
+        <View>
+          <Text style={styles.errorMessage} numberOfLines={3}>
+            {getErrorMessage(error, strings)}
+          </Text>
+        </View>
+      ) : null}
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Swap/common/ConfirmWithSpendingPassword/ConfirmWithSpendingPassword.tsx]:
```suggestion
      {error != null && (
        <View>
          <Text style={styles.errorMessage} numberOfLines={3}>
            {getErrorMessage(error, strings)}
          </Text>
        </View>
      )}
```

File: apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/helpers.ts [hunk 10]
```diff
@@ -0,0 +1,601 @@
...
+    async (options: Options) => {
+      const address = await CardanoMobile.Address.fromBech32(options.bech32Address)
+      const bytes = await address.toBytes()
+      const addressHex = new Buffer(bytes).toString('hex')
+      const cbor = await order.cancel({
+        utxos: {collateral: options.collateralUtxo, order: options.orderUtxo},
+        address: addressHex,
+      })
+      const tx = await CardanoMobile.Transaction.fromBytes(Buffer.from(cbor, 'hex'))
+      const feeNumber = await tx.body().then((b) => b.fee())
```
Comment (lisicky) [inline apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/helpers.ts:38]: Is it just tx fee or it should be fee for a cancel action ?

Comment (michalszorad) [reply apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/helpers.ts:38]: This computes only the TX fee. Are there any other fees associated with canceling a limit order?

File: apps/wallet-mobile/__mocks__/@emurgo/csl-mobile-bridge.js
Comment (michalszorad) [reply apps/wallet-mobile/__mocks__/@emurgo/csl-mobile-bridge.js:1]: Removed

File: apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/signatureUtils.ts
Comment (michalszorad) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/signatureUtils.ts:1]: I've moved this to `common` folder

File: apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/helpers.ts (hunk 10 above)
Comment (lisicky) [reply apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/helpers.ts:38]: I don't think. I was just double checking

File: apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/helpers.ts [hunk 11]
```diff
@@ -0,0 +1,601 @@
...
+    .then(async (builder) => builder.keyDeposit(await bigNumFromStr('2000000')))
+    .then((builder) => builder.maxValueSize(4000))
+    .then((builder) => builder.maxTxSize(8000))
+    .then(async (builder) => builder.coinsPerUtxoWord(await bigNumFromStr('34482')))
+    .then(async (builder) => builder.exUnitPrices(exUnitPrices).then((builder) => builder.build()))
+
+  return CardanoMobile.TransactionBuilder.new(assertRequired(txBuilderCfg, 'Could not build transaction builder'))
+}
+
+const operations1 = Object.values({
```
Comment (lisicky) [inline apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/helpers.ts:109]: I would move it in separate file with all protocol params

File: apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/helpers.ts [hunk 12]
```diff
@@ -0,0 +1,601 @@
...
+  'verifySchnorrSecp256k1Signature-cpu-arguments-slope': 32947,
+  'verifySchnorrSecp256k1Signature-memory-arguments': 10,
+})
+
+export const assertRequired = <T>(value: T | undefined, message: string): T => {
+  if (value === undefined) throw new Error(message)
+  return value
+}
+
+export const fixScriptHash = async (tx: CSL_TYPES.Transaction) => {
```
Comment (lisicky) [inline apps/wallet-mobile/src/features/Swap/useCases/StartSwapScreen/ListOrders/helpers.ts:461]: You can recalculate script hash without building tx by tx builder. There is function `hash_script_data` that you can use

File: apps/wallet-mobile/src/yoroi-wallets/cardano/common/signatureUtils.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
+import {PrivateKey} from '@emurgo/cross-csl-core'
+import blake2b from 'blake2b'
+
+import {CardanoMobile} from '../../wallets'
+
+export const signRawTransaction = async (txHex: string, pKeys: PrivateKey[]): Promise<Uint8Array> => {
+  const fixedTx = await CardanoMobile.FixedTransaction.fromHex(txHex)
```
Comment (lisicky) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/common/signatureUtils.ts:7]: it you have byte array you can call `fromBytes`

**Review by lisicky — APPROVED** [review]

**Pull request merged**